### PR TITLE
fix: refuse stale pivot cache saves

### PR DESCRIPTION
**Pull request by jpbesgen**
## Summary

- resolve worksheet-backed pivot sources through direct ranges, static defined names, and named worksheet tables
- refuse validation/save when recorded value edits would silently stale a preserved pivot cache unless the caller explicitly selects that cache for refresh-on-open
- report the remaining Excel refresh requirement in edit receipts, including when `refreshOnLoad` was already enabled
- preserve unrelated local edits and exclude external pivot sources from local dependency blocking; unresolved local sources refuse conservatively

This is a preservation-safety fix only. It does not add pivot calculation, semantic pivot editing, pivot creation, or a new rendering dependency.

## Validation

- `pytest -q tests/paper/test_delivery.py` — 26 passed
- `pytest -q tests/paper -m 'not lo_smoke'` — 758 passed, 1 skipped, 25 deselected
- `pytest -q openpyxl` — 2,592 passed, 6 skipped, 7 xfailed
- focused upstream pivot, defined-name, and table tests — 163 passed, 2 xfailed
- warning-clean Sphinx build with the CI command
- compile, focused Ruff error/import checks, and `git diff --check`

The local macOS LibreOffice installation still aborts with the checkout's known `rc=134`, including its untouched-fixture self-test. The Linux LibreOffice CI leg remains the authoritative integration check.

---
## Summary by cubic
Refuses saves that would leave a worksheet-backed pivot cache stale and reports that Excel must refresh affected caches on open. Previously these saves kept plausible but outdated pivot results; now validate/save fail unless the caller selects the affected cache for refresh-on-open.

- Detects impacts from direct ranges, static defined names, named worksheet tables, transitive formula inputs (including array/data-table spills), staged formula-cache writes, source retargeting, and formula-binding changes. Treats known volatiles (including `CELL` and `INFO`) as changing on value writes and propagates calculation context changes (cell formatting, row/column visibility, AutoFilter via `SUBTOTAL`/`AGGREGATE`). External sources are ignored; unresolved local sources refuse conservatively. Structural sheet edits use parsed pivot sources.
- Oracle candidates auto-select impacted caches, set refresh metadata, and expose `RecalcResult.pivot_refreshes`. Save receipts include `pivot_source_changed_requires_refresh` with `cause` (`input_changed` or `source_changed`) and `requirement` `excel_refresh_on_open`.
- Refresh planning sets `enableRefresh="1"` and `refreshOnLoad="1"` only when needed; unrelated edits preserve cache bytes. Headless readers may see old cached results until Excel refreshes.
- Performance: skips analysis when no pivots are present and indexes dependency propagation to avoid quadratic scans.

- Required migration: when editing cells that feed a pivot, changing a pivot’s source or bindings, expanding a named table used by a pivot, writing formula caches, or changing calculation context (formatting, row/column visibility, filters), call `wb.set_pivot_refresh_on_load(pivots=[...])` or `wb.set_pivot_refresh_on_load(all=True)` before validate/save. Consent is required even if refresh-on-open was already enabled. Pivots whose sources include a UDF must be explicitly selected for refresh.

<sup>Written for commit 36c0d58b31fb8cfd716041c1b017a67f365b3526. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/paper-instruments/paper-xlsx/pull/20?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

---

> [!NOTE]
> **Medium Risk**
> Changes the preserve-mode save/validate contract for workbooks with local pivots (new refusals and explicit-consent requirement) and adds non-trivial formula-dependency analysis on the save path. Unrelated edits and files without pivots should be unaffected.
> 
> **Overview**
> Preserve-mode validate/save now **refuses edits that would leave a worksheet-backed pivot cache stale**, unless the caller explicitly selects those pivots via `wb.set_pivot_refresh_on_load(...)`. Previously such saves could keep plausible but outdated pivot results.
> 
> Impact detection covers direct ranges, static defined names, named tables, source retargeting, transitive formula inputs (including array/data-table spills and staged cache writes), known volatiles (`NOW`/`TODAY`/`RAND`/`CELL`/`INFO`), and calculation-context changes (formatting, row/column visibility, AutoFilter for `SUBTOTAL`/`AGGREGATE`). Unresolved local sources refuse conservatively; external sources do not block unrelated local edits. Structural sheet edits use parsed pivot sources instead of byte-needle matching.
> 
> Consent still required even if `refreshOnLoad` was already set. Receipts add `pivot_source_changed_requires_refresh` (`excel_refresh_on_open`). Oracle preserved candidates auto-select impacted caches and expose `RecalcResult.pivot_refreshes`. Refresh planning also re-enables `enableRefresh` when it was disabled.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c9f531dc10eb57b6befe4daf1499fce8d3d150c5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: openpyxl/preserve/pivots.py [hunk 1]
```diff
@@ -105,6 +105,269 @@ def descendants(node):
...
+
+
+def _bounds(ref):
+    try:
+        bounds = range_boundaries(ref)
+    except (TypeError, ValueError):
+        return None
+    if any(value is None for value in bounds):
+        return None
+    return bounds
```
[the remark this answers is not in the document]
Comment (jpbesgen) [reply openpyxl/preserve/pivots.py:151]: Fixed in b103712b. Open axes returned by range_boundaries are now retained as exact source bounds instead of being downgraded to unresolved; the existing intersection logic expands only the open axis. Added focused whole-column and whole-row cases proving unrelated edits remain allowed.

File: openpyxl/preserve/pivots.py [hunk 2]
```diff
@@ -105,6 +105,269 @@ def descendants(node):
...
+            ws, bounds = resolved
+            return ws, bounds, name, binding
+        return None, None, "unresolved worksheet pivot source", binding
+    return (None, None, "unresolved worksheet pivot source",
+            ("worksheet-source", title, ref, name))
+
+
+def snapshot_sources(wb):
+    """Capture semantic local-source bindings when preservation arms."""
+    _index_map, cache_parts = _index(wb)
```
[the remark this answers is not in the document]
Comment (jpbesgen) [reply openpyxl/preserve/pivots.py:273]: Fixed in b103712b. Pivot-table roots now receive the same expanded-namespace validation and legal prefixed-root handling as cache definitions. The existing prefixed-pivot regression now covers both prefixed pivotTableDefinition and pivotCacheDefinition parts.

File: openpyxl/preserve/pivots.py [hunk 3]
```diff
@@ -105,6 +105,269 @@ def descendants(node):
...
+        (ws.title, row, column)
+        for ws, coordinates in dirty.items()
+        for row, column in coordinates
+        if row is not None and column is not None
+    }
+    if not tainted:
+        return tainted
+    from .perception import dependency_sketch
+
+    sketch = dependency_sketch(wb)
```
[the remark this answers is not in the document]
Comment (jpbesgen) [reply openpyxl/preserve/pivots.py]: Fixed in b103712b. Data-only pivot checks now build the dependency sketch from the retained source formula view, so an edit to D2 taints a preserved A2 formula even when the live model contains only cached values. Added the focused data-only regression; the broader preservation set passes (57 tests) and the full non-LibreOffice Paper suite passes (781 tests).

File: openpyxl/preserve/pivots.py [hunk 4]
```diff
@@ -105,6 +105,269 @@ def descendants(node):
...
+    # An unresolved formula may read any edited cell. Taint it rather than
+    # guessing that its dynamic/structured/external reference is unrelated.
+    for address in sketch.unresolved:
+        key = _address_key(address)
+        if key is not None:
+            tainted.add(key)
```
[the remark this answers is not in the document]
Comment (jpbesgen) [reply openpyxl/preserve/pivots.py]: Fixed in b103712b. The dependency closure now carries each array or spill anchor output range and uses range intersection for downstream propagation and pivot-source checks, avoiding cell enumeration. Added a regression where the pivot source contains only follower cells.

File: openpyxl/preserve/pivots.py [hunk 5]
```diff
@@ -105,6 +123,542 @@ def descendants(node):
...
+
+
+def source_impacts(wb, ledger):
+    """Return local pivot caches that workbook edits can make stale."""
+    index, _cache_parts = _index(wb)
+    if not index:
+        return []
+    changes = {ws: set(coords)
+               for ws, coords in ledger.value_overwrites.items() if coords}
+    for ws, writes in ledger.cache_writes.items():
```
Comment (daanishkhazi) [inline openpyxl/preserve/pivots.py:832]: **P1: Preserve the oracle output path for formula-backed pivot sources**

`oracle._preserved_recalc_candidate()` loads a private preserve-mode workbook, records fresh LibreOffice values in `ledger.cache_writes`, then calls `wb_candidate.save(buf)`. This loop now treats those internal cache writes as source edits, but the private workbook has no `pivot_refresh_requests`, and `oracle.recalc(..., output_path=...)` exposes no way for its caller to add one. I reproduced the public workflow: recalculating a formula cell inside a pivot source makes the candidate save raise `stale-pivot-cache`.

Please integrate the oracle explicitly rather than bypassing this validation. Affected pivot caches should be selected under a documented oracle refresh policy, or the caller should receive a bounded option that the internal candidate can apply. The result should still disclose that Excel must refresh the pivot. Please add an integration regression through `oracle.recalc(..., output_path=...)`, not only a direct `ledger.cache_writes` test.

File: openpyxl/preserve/pivots.py [hunk 6]
```diff
@@ -105,6 +123,542 @@ def descendants(node):
...
+        before = len(tainted_ranges)
+        for key in tainted:
+            output = outputs.get(key)
+            if output is not None:
+                tainted_ranges.add(output)
+        return len(tainted_ranges) != before
+
+    # An unresolved formula may read any edited cell. Taint it rather than
+    # guessing that its dynamic/structured/external reference is unrelated.
+    for address in sketch.unresolved:
```
Comment (daanishkhazi) [inline openpyxl/preserve/pivots.py:803]: **P1: Taint volatile pivot-source formulas**

The freshness analysis still misses known volatile formulas. `NOW()`, `TODAY()`, `RAND()`, `RANDBETWEEN()`, and `RANDARRAY()` have no `RANGE` operands and are not placed in `dependency_sketch().unresolved`, so this closure never taints their outputs. I reproduced a pivot over `A1 = NOW()`: after editing an unrelated value cell, `wb.validate()` succeeds without refresh consent. Excel may then recalculate `A1` on open while retaining the old pivot cache, recreating the silent-staleness mode this PR is intended to prevent.

The perception module already defines the volatile function set. Please classify those formula outputs as potentially changed for a value-editing save, or otherwise force the affected pivot refresh contract, and add focused `NOW`/`TODAY`/`RAND` regressions. Unknown zero-argument UDFs should also have an explicit conservative or documented boundary.

**Review by daanishkhazi — COMMENTED** [review]
I found two correctness gaps that should be resolved before release because they undermine the stale-pivot-cache contract.

**Comment by daanishkhazi** [comment]
## Performance blocker: make pivot dependency propagation indexed and work-queue based

I re-tested the latest head after the oracle, volatility, calculation-context, and AutoFilter fixes. Those fixes look directionally correct, and the focused/full Paper suites pass. I think one material issue remains before merge: `_dirty_closure()` has pathological propagation cost on a common formula-backed PivotTable shape.

### This is not only an adversarial deep-chain concern

A realistic example is a transaction table with one helper formula per row:

```text
Data!Z1                         = shared FX/normalization assumption
Data!B2:B50001                 = raw transaction amounts
Data!A2                        = $Z$1 * B2
Data!A3                        = $Z$1 * B3
...
Data!A50001                    = $Z$1 * B50001
Pivot source                   = Data!A1:B50001
Edit                           = Z1 changes from 1.0 to 1.1
```

This is a normal revenue, GL, customer-concentration, transaction-diligence, or operating-data workbook: a banker changes an FX rate, classification mapping, selected-case factor, or normalization assumption, and thousands of helper values feeding a pivot must refresh.

I generated this shape and timed only `wb.validate()` after the workbook was already loaded:

| Workbook | Validation time |
|---|---:|
| 20,000 formulas, no pivot | 0.719 s |
| 20,000 formulas, local pivot over the helper column | 1.956 s |
| 50,000 formulas, no pivot | 1.855 s |
| 50,000 formulas, local pivot over the helper column | **32.144 s** |

The same effect is visible in a reverse-ordered dependency chain: 200 formulas took 0.268 s, 500 took 2.799 s, and 1,000 took 17.237 s. That chain is deliberately adversarial, but the 50,000-row fan-out is a practical PivotTable workload.

The cost can also be paid repeatedly. `validate()` performs the analysis, `save()` validates again, and a receipt-producing save can call `source_impacts()` again while deriving pivot effects. Explicitly selecting the affected pivot for refresh does not remove the analysis because the save still needs the impact list and receipt evidence.

### Exact cause

The current closure is structurally equivalent to:

```python
tainted = initial_dirty_cells
changed = True
while changed:
    changed = False
    for formula, references in every_formula:
        if formula already in tainted:
            continue
        if any(reference intersects any cell/range in tainted):
            tainted.add(formula)
            changed = True
```

There are two compounding costs:

1. The `while changed` loop can rescan every formula once per dependency level. Formula iteration follows worksheet/cell order, not topological order, so a reverse-ordered chain can advance only one link per complete pass.
2. `_bounds_hit()` and `_ranges_hit()` linearly scan the entire growing `tainted`/`tainted_ranges` collections for each reference. Even a one-level fan-out can therefore become superlinear as each discovered helper formula enlarges the set subsequently scanned by every remaining formula.

`dependency_sketch()` itself is not the regression; a single formula-tokenization pass is expected. The repeated fixed-point traversal and linear intersection checks are the new cost introduced by this pivot-freshness path.

### Proposed fix

Please preserve the current safety semantics and replace only the traversal strategy. We should not solve this by refreshing every pivot, refusing every edit in a pivot workbook, limiting dependency depth, or checking only direct source writes. Those alternatives either over-mutate, over-refuse, or recreate the stale-cache hole.

Build a reverse dependency index once from the existing `DependencySketch`, then propagate through a work queue:

```text
initial dirty cells/ranges
    -> indexed formulas that read those cells/ranges
    -> newly tainted formula output cells/ranges
    -> indexed formulas that read those outputs
    -> repeat until the queue is empty
```

A concrete private representation could be:

```python
exact_dependents[(sheet_casefold, row, col)] -> set[formula_address]
range_dependents[(sheet_casefold, closed_bounds)] -> set[formula_address]
formula_outputs[formula_address] -> anchor cell and optional array/data-table range
processed_formulas -> set[formula_address]
pending_cells / pending_ranges -> deque
```

Suggested construction and propagation:

1. Parse formulas once with the existing `dependency_sketch()`.
2. Normalize each reference to the existing closed bounds.
3. Put single-cell references in `exact_dependents` for O(1) lookup.
4. Deduplicate identical rectangular references before indexing them. This matters for thousands of helper formulas that all reference the same assumption or mapping range.
5. Keep larger/open rectangular references in a per-sheet range index. A simple first implementation can index unique ranges by sheet; it does not need a new dependency. If the number of unique ranges becomes large, bucket them by row bands or use a small internal interval index.
6. Seed the queue with current direct value/cache writes and calculation-context cells/ranges.
7. Seed unresolved and known-volatile formulas exactly as today. Their conservative behavior should not change.
8. When a formula is discovered, mark it once, add its anchor output to the tainted set, and enqueue its declared array/data-table output range when present.
9. Query dependents only for newly tainted frontier items. Never rescan a formula after it is in `processed_formulas`.
10. Preserve the current special handling for contextual formulas (`SUBTOTAL`/`AGGREGATE`), style/display/filter activation, binding changes, and `force_recalculation`.

The desired complexity is approximately `O(formulas + dependency edges + range-query cost)`. Exact-cell fan-out should be linear. Each formula should become tainted at most once regardless of worksheet order.

A smaller interim optimization such as checking only the newly tainted frontier would reduce the worst behavior, but without a reverse index it still scans every formula once per dependency depth. Since the index can stay private to `pivots.py`, I recommend fixing the underlying traversal rather than layering another partial optimization onto this PR.

### Correctness requirements

The optimized closure must produce the same `tainted` cells and ranges as the current fixed-point implementation for all supported/conservative classes:

- direct and transitive references;
- whole-row and whole-column references;
- static names and table operands as currently classified;
- unresolved/dynamic references;
- known volatile formulas;
- array and data-table output ranges;
- formula-binding changes;
- style, row/column, filter, and contextual-function changes;
- `data_only=True` retained-formula analysis;
- oracle cache writes and forced recalculation.

A useful way to de-risk the rewrite is to retain the current closure as a test-only reference implementation and compare optimized vs. reference results on small generated dependency graphs. That tests semantic equivalence without adding broad defensive surface to production code.

### Focused tests / acceptance criteria

I would add only tests that establish the algorithmic boundary:

1. **High fan-out:** thousands of formula cells all reference one edited assumption and feed the pivot source. Every output is discovered exactly once.
2. **Reverse chain:** dependencies are ordered opposite worksheet traversal. The result reaches the pivot source without repeated full-graph passes.
3. **Shared range:** many formulas reference one common rectangular mapping range; the range is indexed/deduplicated and all dependents are found.
4. **Array output:** tainting an anchor enqueues the declared output range and reaches a pivot over follower cells.
5. **Semantic parity:** optimized and reference closures match on small combinations of exact, ranged, unresolved, volatile, and contextual dependencies.

Avoid a brittle strict wall-clock assertion in CI. Prefer an instrumentable traversal count, such as asserting each formula is processed at most once and that intersection-query counts remain proportional to formulas/unique ranges. As an external benchmark, the 50,000-row fan-out should land near the roughly two-second no-pivot parsing/validation baseline rather than 32 seconds.

This is confined to the new pivot-freshness analysis. Workbooks without pivots already return before the closure, and that behavior should remain unchanged. I recommend treating this as the final performance blocker for PR #20 rather than reducing any of the safety coverage Joey added.

**Review by daanishkhazi — APPROVED** [review]
